### PR TITLE
Decimal improvements

### PR DESCRIPTION
**Pull request by 4ertus2**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

### CODE REVIEW

File: dbms/src/Functions/FunctionsHashing.h [hunk 1]
```diff
@@ -285,7 +291,7 @@ class FunctionIntHash : public IFunction
 
     DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
     {
-        if (!arguments[0]->isValueRepresentedByNumber())
+        if (!allowIntHash(arguments[0].get()))
```
Comment (alexey-milovidov) [inline dbms/src/Functions/FunctionsHashing.h]: It's all right to hash decimal exactly as the underlying integer.

File: dbms/src/DataTypes/DataTypesDecimal.h [hunk 2]
```diff
@@ -131,23 +131,14 @@ class DataTypeDecimal final : public DataTypeSimpleSerialization
     bool isParametric() const override { return true; }
     bool haveSubtypes() const override { return false; }
     bool shouldAlignRightInPrettyFormats() const override { return true; }
-    bool textCanContainOnlyValidUTF8() const override { return true; }
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypesDecimal.h]: I cannot find the replacement.

textCanContainOnlyValidUTF8 - that is true for decimal.

File: dbms/src/DataTypes/DataTypesDecimal.h [hunk 3]
```diff
@@ -134,20 +134,12 @@ class DataTypeDecimal final : public DataTypeSimpleSerialization
...
     size_t getSizeOfValueInMemory() const override { return sizeof(T); }
-    bool isCategorial() const override { return isValueRepresentedByInteger(); }
 
-    bool canBeUsedAsVersion() const override { return false; }
     bool isSummable() const override { return true; }
-    bool canBeUsedInBitOperations() const override { return false; }
-    bool isUnsignedInteger() const override { return false; }
     bool canBeUsedInBooleanContext() const override { return true; }
-    bool isNumber() const override { return true; }
-    bool isInteger() const override { return false; }
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/DataTypesDecimal.h]: Why not?

File: dbms/src/DataTypes/IDataType.h [hunk 4]
```diff
@@ -423,5 +423,77 @@ class IDataType : private boost::noncopyable
...
+    bool isDateTime() const { return idx == TypeIndex::DateTime; }
+    bool isDateOrDateTime() const { return isDate() || isDateTime(); }
+
+    bool isString() const { return idx == TypeIndex::String; }
+    bool isFixedString() const { return idx == TypeIndex::FixedString; }
+    bool isStringOrFixedString() const { return isString() || isFixedString(); }
+
+    bool isUUID() const { return idx == TypeIndex::UUID; }
+    bool isArray() const { return idx == TypeIndex::Array; }
+    bool isTuple() const { return idx == TypeIndex::Tuple; }
```
Comment (alexey-milovidov) [inline dbms/src/DataTypes/IDataType.h]: Now you have to replace typeid_casts, checkDataType across all the codebase.

Comment (4ertus2) [reply dbms/src/DataTypes/IDataType.h]: OK

**Pull request merged**